wa/framework/workload.py

```diff
@@ -865,3 +865,42 @@ def teardown(self):
         self.target.execute('am force-stop {}'.format(self.apk_info.package))
         if self.uninstall:
             self.target.uninstall_package(self.apk_info.package)
+
+class TestPackageHandler(PackageHandler):
+    """Class wrapping an APK used through ``am instrument``.
+    """
+    def __init__(self, owner, instrument_args=None, raw_output=False,
+                 instrument_wait=True, no_hidden_api_checks=False,
+                 *args, **kwargs):
+        if instrument_args is None:
+            instrument_args = {}
+        super(TestPackageHandler, self).__init__(owner, *args, **kwargs)
+        self.raw = raw_output
+        self.args = instrument_args
+        self.wait = instrument_wait
+        self.no_checks = no_hidden_api_checks
+        self.output = None
+        self.cmd = ''
+
+    def setup(self, context):
+        self.initialize_package(context)
+        options = ' '.join('-e {} {}'.format(k, v)
+                           for k, v in self.args.items())
+        r_flag = '-r' if self.raw else ''
+        w_flag = '-w' if self.wait else ''
+        n_flag = '--no-hidden-api-checks' if self.no_checks else ''
+        flags = ' '.join(filter(None, [w_flag, r_flag, n_flag, options]))
+
+        target = (self.apk_info.package if not self.apk_info.activity else
+                  '{}/{}'.format(self.apk_info.package, self.apk_info.activity))
+        self.cmd = 'am instrument {} {}'.format(flags, target)
+
+    def start_activity(self):
+        """NB: Unlike regular applications, this is a blocking call."""
```

Comment:
This breaks the substitution principle. Ether:

- Make this non-blocking by kicking it of  in a thread internally ant returning right away (and then prividing `wait()` or similar), or
-  Create a common base for this and `PackageHandler`, rather than subclassing `PackageHandler`, and implement a `run_activity()` rather than `start_activity()`.

Reply:
This difference in behaviour comes from the difference between `am start` (non-blocking) and `am instrument` (blocking) and my attempt at adding support for `am instrument`-based APKs in a somewhat generic yet lightweight way (in terms of required change).

I agree that the best way would be to rethink the class inheritance tree to properly integrate that distinction while keeping polymorphism but that's more than I'm ready to take up right now. I've implemented the non-blocking approach and the workload's `run` method starts the activity then waits for the instrument to return, while the `TestPackageHandler.start_activity` remains non-blocking (as it creates a thread on the host).